### ISSUE TITLE
image/jpeg responses cached as application/json

### ISSUE THREAD

**Issue by petermorlion** [filipw/Strathweb.CacheOutput#219]
I'm trying to cache `image/jpeg` responses, using Redis, but noticed the data is stored as `application/json; charset=utf-8`. So when the response is taken from the cache, I get an invalid response.

My code is doing something like:

```
response = new HttpResponseMessage(HttpStatusCode.OK);
response.Content = new StreamContent(memoryStream);
response.Content.Headers.ContentType = new MediaTypeHeaderValue("image/jpeg");
return response;
```

But I see in the Redis cache, that there is a key `application/json`.

I think that the `GetExpectedMediaType` method will always return `application/json`, even though my response was `image/jpeg`. This is because the client won't always send `Accept:image/jpeg` in the headers and/or because there isn't a formatter for `image/jpeg` (these conditions are checked [here](https://github.com/filipw/Strathweb.CacheOutput/blob/37bc9323259838e8fd2eb44f6bbde3251bd822ee/src/WebApi.OutputCache.V2/CacheOutputAttribute.cs#L149)).

So I get the impression that CacheOutput doesn't seem to cache the output like I originally created it. Is this considered a bug or is it by design and should I subclass the attribute (like you suggested in #52)?

**Comment by filipw** [comment]
Very good question. The library will use the content negotiation of Web API to determine the output response media type and fallback to JSON.
Since in your case you bypass conneg completely, indeed you get weird results.

The library relies on symmetry between requests and responses, meaning it assumes the media type can be computed both from the request and from the response.

I think the simplest solution would be to update the library [here](https://github.com/filipw/Strathweb.CacheOutput/blob/37bc9323259838e8fd2eb44f6bbde3251bd822ee/src/WebApi.OutputCache.V2/CacheOutputAttribute.cs#L223) and look at the content type of the response, in case it was manually set in the action like your case. This still means that the caller will need to send `image/jpeg` as the `Accept` header in order for the cache match to occur successfully.

The alternative, more robust solution, would be to explicitly allow overriding media type from the caching attribute - since in your case you know that you will only be caching one specific media type. Something like `[CacheOutput(MediaType = "image/jpeg")]`

**Comment by petermorlion** [comment]
If you want to go the attribute-route, I'd be willing to submit a PR, unless you want to look into it yourself?

**Comment by filipw** [comment]
> If you want to go the attribute-route, I'd be willing to submit a PR

that would be awesome!

**Comment by petermorlion** [comment]
See #220 Happy to hear what you think about it.

**Comment by filipw** [comment]
Fixed by #220
